### <p style="text-align: center;"> **Name:** Shahzaib Khan &ensp;&ensp;&ensp;&ensp;&ensp;&ensp; **Roll No:** 19K-0273 &ensp;&ensp;&ensp;&ensp;&ensp;&ensp; **Section:** BCS-8B</p>

## **Question: 2**
Cleaning the Datasets in the _'data'_ folder using Numpy and Pandas

In [9]:
import numpy as np, pandas as pd

### **Part 'A':** 'BL_Flickr_Images_Book.csv' Dataset
- [Dropping Columns in a DataFrame](<#####2)-Dropping-unnecessary-columns>)
- [Changing the Index of a DataFrame](<#####3)-Set-'Identifier'-column-as-index>)
- [Tidying up Fields in the Data](<#####4)-Cleaning-fields-in-the-data-(dtypes-and-data-inconsistencies)>)
- [Combining 'str' Methods with NumPy to Clean Columns](<######toc-reference>)

##### 1) Load the data

In [44]:
df_book = pd.read_csv("data/bl_flickr_images_book.csv")
df_book.head()

,Identifier,Edition Statement,Place of Publication,Date of Publication,Publisher,Title,Author,Contributors,Corporate Author,Corporate Contributors,Former owner,Engraver,Issuance type,Flickr URL,Shelfmarks
0,206,NaN,London,1879 [1878],S. Tinsley & Co.,Walter Forbes. [A novel.] By A. A,A. A.,"FORBES, Walter.",NaN,NaN,NaN,NaN,monographic,http://www.flickr.com/photos/britishlibrary/ta...,British Library HMNTS 12641.b.30.
1,216,NaN,London; Virtue & Yorston,1868,Virtue & Co.,All for Greed. [A novel. The dedication signed...,"A., A. A.","BLAZE DE BURY, Marie Pauline Rose - Baroness",NaN,NaN,NaN,NaN,monographic,http://www.flickr.com/photos/britishlibrary/ta...,British Library HMNTS 12626.cc.2.
2,218,NaN,London,1869,"Bradbury, Evans & Co.",Love the Avenger. By the author of “All for Gr...,"A., A. A.","BLAZE DE BURY, Marie Pauline Rose - Baroness",NaN,NaN,NaN,NaN,monographic,http://www.flickr.com/photos/britishlibrary/ta...,British Library HMNTS 12625.dd.1.
3,472,NaN,London,1851,James Darling,"Welsh Sketches, chiefly ecclesiastical, to the...","A., E. S.","Appleyard, Ernest Silvanus.",NaN,NaN,NaN,NaN,monographic,http://www.flickr.com/photos/britishlibrary/ta...,British Library HMNTS 10369.bbb.15.
4,480,"A new edition, revised, etc.",London,1857,Wertheim & Macintosh,"[The World in which I live, and my place in it...","A., E. S.","BROOME, John Henry.",NaN,NaN,NaN,NaN,monographic,http://www.flickr.com/photos/britishlibrary/ta...,British Library HMNTS 9007.d.28.


##### 2) Dropping unnecessary columns

We can see some columns provide ancillary information that would be useful
for the library but aren't very descriptive of the books.
Drop those columns to have just the information about the books.

In [45]:
# list of columns to be dropped
to_drop = [
    'Edition Statement', 'Corporate Author', 'Corporate Contributors',
    'Former owner', 'Engraver', 'Contributors', 'Issuance type',
    'Shelfmarks'
]

df_book.drop(columns=to_drop, inplace=True)
df_book.head()

,Identifier,Place of Publication,Date of Publication,Publisher,Title,Author,Flickr URL
0,206,London,1879 [1878],S. Tinsley & Co.,Walter Forbes. [A novel.] By A. A,A. A.,http://www.flickr.com/photos/britishlibrary/ta...
1,216,London; Virtue & Yorston,1868,Virtue & Co.,All for Greed. [A novel. The dedication signed...,"A., A. A.",http://www.flickr.com/photos/britishlibrary/ta...
2,218,London,1869,"Bradbury, Evans & Co.",Love the Avenger. By the author of “All for Gr...,"A., A. A.",http://www.flickr.com/photos/britishlibrary/ta...
3,472,London,1851,James Darling,"Welsh Sketches, chiefly ecclesiastical, to the...","A., E. S.",http://www.flickr.com/photos/britishlibrary/ta...
4,480,London,1857,Wertheim & Macintosh,"[The World in which I live, and my place in it...","A., E. S.",http://www.flickr.com/photos/britishlibrary/ta...


##### 3) Set 'Identifier' column as index

First of all, check if the column is unique or not.

In [46]:
df_book['Identifier'].is_unique

True

In order to make sure it would be a good idea to use this column as index,
let's see at some example.

In [47]:
df_book.loc[0]

Identifier                                                            206
Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 0, dtype: object

It can be noticed that the 1st record in our dataframe has the identifier **206** which can be accessed in a straightforward way with `loc[]`.

`loc[]` allows us to to label-based indexing (_labeling of a row regardless of
its position_), so why not use the 'Identifier' column as index so it's easier
just to look up all the records using their index which **after update** will
be the same as their identifier.

In [48]:
df_book.set_index('Identifier', inplace=True)
df_book.loc[206]

Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

We notice that the index of our **1st record** has been updated from being **0 to 206** which is the same as its identifier.

##### 4) Cleaning fields in the data (dtypes and data inconsistencies)

It's important to get specific columns like date to a uniform format for
better understanding and enforce consistency.

Let's first check the different datatypes we have in our dataset.

In [49]:
df_book.dtypes.value_counts() # [NOTE]: DataFrame.get_dtype_counts() is deprecated!

object    6
Name: count, dtype: int64

In [50]:
df_book.dtypes

Place of Publication    object
Date of Publication     object
Publisher               object
Title                   object
Author                  object
Flickr URL              object
dtype: object

You can see that all of the datatypes are currently **'object'** in native Python.

**'Date of Publication'** is one column where it makes sense to enforce a **numeric
value** so calculations can be performed later on.

In [52]:
# let's check a few values first
df_book['Date of Publication'].head(15)

Identifier
206     1879 [1878]
216            1868
218            1869
472            1851
480            1857
481            1875
519            1872
667             NaN
874            1676
1143           1679
1280           1802
1808           1859
1905           1888
1929    1839, 38-54
2836           1897
Name: Date of Publication, dtype: object

We can notice a few issues here:
1. **extra dates in []:** 1879 [1868]
2. **date ranges:** 1839, 38-54
3. **NaN**s

We can fix these with **regex** as we notice that we only need the **first 4 digits** to get our year correct.

In [53]:
regex = r'^(\d{4})' # to find any 4 digits at the beginning
date_num = df_book['Date of Publication'].str.extract(regex, expand=False)
date_num.head(15)

Identifier
206     1879
216     1868
218     1869
472     1851
480     1857
481     1875
519     1872
667      NaN
874     1676
1143    1679
1280    1802
1808    1859
1905    1888
1929    1839
2836    1897
Name: Date of Publication, dtype: object

Now that we've got our desired year values, the **dtype** of this column is still
**'object'**. let's make it **numeric**.

In [54]:
df_book['Date of Publication'] = pd.to_numeric(date_num, downcast='integer')
df_book['Date of Publication'].dtype

dtype('float64')

We still have **NaN**s, let's check how many we have:

In [55]:
no_of_null_values = int(df_book['Date of Publication'].isnull().sum())
no_of_null_values # 971

971

Maybe a **percentage** over the whole set is better to decide if it's problematic:

In [56]:
no_of_null_values / len(df_book) * 100

11.717147339205985

**~11.72%** of our data is **missing** which is not a big problem for us right now.
but still we can just skip those records:

In [57]:
df_book = df_book[df_book['Date of Publication'].isnull() != True]
df_book.head(15)

,Place of Publication,Date of Publication,Publisher,Title,Author,Flickr URL
Identifier,,,,,,
206,London,1879.0,S. Tinsley & Co.,Walter Forbes. [A novel.] By A. A,A. A.,http://www.flickr.com/photos/britishlibrary/ta...
216,London; Virtue & Yorston,1868.0,Virtue & Co.,All for Greed. [A novel. The dedication signed...,"A., A. A.",http://www.flickr.com/photos/britishlibrary/ta...
218,London,1869.0,"Bradbury, Evans & Co.",Love the Avenger. By the author of “All for Gr...,"A., A. A.",http://www.flickr.com/photos/britishlibrary/ta...
472,London,1851.0,James Darling,"Welsh Sketches, chiefly ecclesiastical, to the...","A., E. S.",http://www.flickr.com/photos/britishlibrary/ta...
480,London,1857.0,Wertheim & Macintosh,"[The World in which I live, and my place in it...","A., E. S.",http://www.flickr.com/photos/britishlibrary/ta...
481,London,1875.0,William Macintosh,"[The World in which I live, and my place in it...","A., E. S.",http://www.flickr.com/photos/britishlibrary/ta...
519,London,1872.0,The Author,Lagonells. By the author of Darmayne (F. E. A....,"A., F. E.",http://www.flickr.com/photos/britishlibrary/ta...
874,London],1676.0,NaN,"A Warning to the inhabitants of England, and L...",Remaʿ.,http://www.flickr.com/photos/britishlibrary/ta...
1143,London,1679.0,NaN,A Satyr against Vertue. (A poem: supposed to b...,"A., T.",http://www.flickr.com/photos/britishlibrary/ta...


Now let's check a (_possible_) **string** column **'Place of Publication'** for inconsistencies:

In [58]:
df_book['Place of Publication'].head(15)

Identifier
206                       London
216     London; Virtue & Yorston
218                       London
472                       London
480                       London
481                       London
519                       London
874                      London]
1143                      London
1280                    Coventry
1808                 Christiania
1905                     Firenze
1929                   Amsterdam
2836                      Savona
2854                      London
Name: Place of Publication, dtype: object

We see that for some rows, the **'place of publications'** contains some unnecessary
information. further check leads us to identify **'London'** and **'Oxford'** to be
the values that we are actually concerned with.

In [ ]:
# comparing 2 entries for another issue
book_indexes = [4157862, 4159587]
df_book.loc[book_indexes]

,Place of Publication,Date of Publication,Publisher,Title,Author,Flickr URL
Identifier,,,,,,
4157862,Newcastle-upon-Tyne,1867.0,T. Fordyce,"Local Records; or, Historical Register of rema...","FORDYCE, T. - Printer, of Newcastle-upon-Tyne",http://www.flickr.com/photos/britishlibrary/ta...
4159587,Newcastle upon Tyne,1834.0,Mackenzie & Dent,"An historical, topographical and descriptive v...","Mackenzie, E. (Eneas)",http://www.flickr.com/photos/britishlibrary/ta...


Here, we see that both values are the same but one of them is separated using **'-'**.

Let's fix all the records for the aforementioned cases with **str** methods combined with **numpy**:

###### _toc reference_

In [ ]:
# combining 'str' methods with numpy to clean columns
df_book['Place of Publication'] = np.where(
    df_book['Place of Publication'].str.contains('London'),
    'London',
    df_book['Place of Publication'].str.replace('-', ' ')
)

Let's test specific examples that we know earlier were to be fixed:

In [61]:
book_indexes = [216, 4157862, 4159587]
df_book.loc[book_indexes]

,Place of Publication,Date of Publication,Publisher,Title,Author,Flickr URL
Identifier,,,,,,
216,London,1868.0,Virtue & Co.,All for Greed. [A novel. The dedication signed...,"A., A. A.",http://www.flickr.com/photos/britishlibrary/ta...
4157862,Newcastle upon Tyne,1867.0,T. Fordyce,"Local Records; or, Historical Register of rema...","FORDYCE, T. - Printer, of Newcastle-upon-Tyne",http://www.flickr.com/photos/britishlibrary/ta...
4159587,Newcastle upon Tyne,1834.0,Mackenzie & Dent,"An historical, topographical and descriptive v...","Mackenzie, E. (Eneas)",http://www.flickr.com/photos/britishlibrary/ta...


let's check the values again:

In [63]:
df_book['Place of Publication'].head(15)

Identifier
206          London
216          London
218          London
472          London
480          London
481          London
519          London
874          London
1143         London
1280       Coventry
1808    Christiania
1905        Firenze
1929      Amsterdam
2836         Savona
2854         London
Name: Place of Publication, dtype: object

And that's all for this dataset.

### **Part 'B':** 'University_Towns.txt' Dataset
- [Cleaning the Entire Dataset Using the applymap Function](######toc-reference-2)

Let’s look at an example. We will create a DataFrame out of the **'university_towns.txt'** file:

In [10]:
%%bash
head data/university_towns.txt

Alabama[edit]
Auburn (Auburn University)[1]
Florence (University of North Alabama)
Jacksonville (Jacksonville State University)[2]
Livingston (University of West Alabama)[2]
Montevallo (University of Montevallo)[2]
Troy (Troy University)[2]
Tuscaloosa (University of Alabama, Stillman College, Shelton State)[3][4]
Tuskegee (Tuskegee University)[5]
Alaska[edit]


Taking advantage of the structure of the file to extract the data into a list of **'(state, city)'** tuples:

In [ ]:
university_towns = []

with open('data/university_towns.txt') as file:
    for line in file:
        if '[edit]' in line:
            # Remember this `state` until the next is found
            state = line
        else:
            # Otherwise, we have a city; keep `state` as last-seen
            university_towns.append((state, line))

university_towns[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

Now we'll convert the list into a DataFrame:

In [ ]:
df_towns = pd.DataFrame(university_towns, columns=['State', 'RegionName'])
df_towns.head()

,State,RegionName
0,Alabama[edit]\n,Auburn (Auburn University)[1]\n
1,Alabama[edit]\n,Florence (University of North Alabama)\n
2,Alabama[edit]\n,Jacksonville (Jacksonville State University)[2]\n
3,Alabama[edit]\n,Livingston (University of West Alabama)[2]\n
4,Alabama[edit]\n,Montevallo (University of Montevallo)[2]\n


Instead of cleaning these strings in **for loop** with `.str` methods, we can use the **`.map`** method to apply a function element-wise in a DataFrame.

###### _toc reference 2_

In [ ]:
def get_citystate(item):
    if ' (' in item:
        return item[:item.find(' (')]
    elif '[' in item:
        return item[:item.find('[')]
    else:
        return item

df_towns = df_towns.map(get_citystate) # [NOTE]: DataFrame.applymap() is deprecated!
df_towns.head()

,State,RegionName
0,Alabama,Auburn
1,Alabama,Florence
2,Alabama,Jacksonville
3,Alabama,Livingston
4,Alabama,Montevallo


**Note:** while it's a convenient and versatile method, `.map` can have significant runtime for larger datasets, because it maps a Python callable to each individual element. In some cases, it can be more efficient to do _`vectorized`_ operations that utilize _`Cython`_ or _`NumPY`_ (which, in turn, makes calls in _`C`_ language) under the hood.

### **Part 'C':** 'Olympics.csv' Dataset
- [Renaming Columns and Skipping Rows](######toc-reference-3)

In [17]:
%%bash
head data/olympics.csv

0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
,? Summer,01 !,02 !,03 !,Total,? Winter,01 !,02 !,03 !,Total,? Games,01 !,02 !,03 !,Combined total
Afghanistan (AFG),13,0,0,2,2,0,0,0,0,0,13,0,0,2,2
Algeria (ALG),12,5,2,8,15,3,0,0,0,0,15,5,2,8,15
Argentina (ARG),23,18,24,28,70,18,0,0,0,0,41,18,24,28,70
Armenia (ARM),5,1,2,9,12,6,0,0,0,0,11,1,2,9,12
Australasia (ANZ) [ANZ],2,3,4,5,12,0,0,0,0,0,2,3,4,5,12
Australia (AUS) [AUS] [Z],25,139,152,177,468,18,5,3,4,12,43,144,155,181,480
Austria (AUT),26,18,33,35,86,22,59,78,81,218,48,77,111,116,304
Azerbaijan (AZE),5,6,5,15,26,5,0,0,0,0,10,6,5,15,26


In [19]:
df_olympics = pd.read_csv('data/olympics.csv')
df_olympics.head()

,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
0,NaN,? Summer,01 !,02 !,03 !,Total,? Winter,01 !,02 !,03 !,Total,? Games,01 !,02 !,03 !,Combined total
1,Afghanistan (AFG),13,0,0,2,2,0,0,0,0,0,13,0,0,2,2
2,Algeria (ALG),12,5,2,8,15,3,0,0,0,0,15,5,2,8,15
3,Argentina (ARG),23,18,24,28,70,18,0,0,0,0,41,18,24,28,70
4,Armenia (ARM),5,1,2,9,12,6,0,0,0,0,11,1,2,9,12


this dataset is a bit messy, lets re-load it with the appropriate parameters:

###### _toc reference 3_

In [ ]:
# let's skip the first row
df_olympics = pd.read_csv('data/olympics.csv', header=1)
df_olympics.head()

,Unnamed: 0,? Summer,01 !,02 !,03 !,Total,? Winter,01 !.1,02 !.1,03 !.1,Total.1,? Games,01 !.2,02 !.2,03 !.2,Combined total
0,Afghanistan (AFG),13,0,0,2,2,0,0,0,0,0,13,0,0,2,2
1,Algeria (ALG),12,5,2,8,15,3,0,0,0,0,15,5,2,8,15
2,Argentina (ARG),23,18,24,28,70,18,0,0,0,0,41,18,24,28,70
3,Armenia (ARM),5,1,2,9,12,6,0,0,0,0,11,1,2,9,12
4,Australasia (ANZ) [ANZ],2,3,4,5,12,0,0,0,0,0,2,3,4,5,12


As we can see, the dataset has some issues. In order to make it readable, we'll label the columns appropriately using the `.rename` method.

In [ ]:
# let's define a dictionary that maps old column names to more usable ones
new_names = {
    'Unnamed: 0': 'Country',
    '? Summer': 'Summer Olympics',
    '01 !': 'Gold',
    '02 !': 'Silver',
    '03 !': 'Bronze',
    '? Winter': 'Winter Olympics',
    '01 !.1': 'Gold.1',
    '02 !.1': 'Silver.1',
    '03 !.1': 'Bronze.1',
    '? Games': '# Games',
    '01 !.2': 'Gold.2',
    '02 !.2': 'Silver.2',
    '03 !.2': 'Bronze.2'
}

df_olympics.rename(columns=new_names, inplace=True)
df_olympics.head()

,Country,Summer Olympics,Gold,Silver,Bronze,Total,Winter Olympics,Gold.1,Silver.1,Bronze.1,Total.1,# Games,Gold.2,Silver.2,Bronze.2,Combined total
0,Afghanistan (AFG),13,0,0,2,2,0,0,0,0,0,13,0,0,2,2
1,Algeria (ALG),12,5,2,8,15,3,0,0,0,0,15,5,2,8,15
2,Argentina (ARG),23,18,24,28,70,18,0,0,0,0,41,18,24,28,70
3,Armenia (ARM),5,1,2,9,12,6,0,0,0,0,11,1,2,9,12
4,Australasia (ANZ) [ANZ],2,3,4,5,12,0,0,0,0,0,2,3,4,5,12


## **Reference:**
https://realpython.com/python-data-cleaning-numpy-pandas/